### DISCUSSION TITLE
Rename Folder Method - GraphException: 403 Access Denied

### DISCUSSION THREAD

**Discussion by whobday** [pnp/pnpcore#1439] [Q&A]
I'm attempting to use the Rename Folder method, but the Graph API is throwing a PnP.Core.MicrosoftGraphServiceException with a HTTPResponseCode of 403 - Access Denied.  

Code:

```
IFolder folder = await pnpContext.Web.GetFolderByServerRelativeUrlAsync(serverRelativeUrl);
await folder.RenameAsync(newName);
```

From fiddler:

```
PATCH https://graph.microsoft.com/v1.0/sites/tmtsp.sharepoint.com,aaaaaaaaaa-7d2b-4d29-b8c4-7f2a446dd277,aaaaaaaa-727b-40c1-8262-d3d720561ed1/drives/b!x4cL5ytaaaaxH8qRG3Sd3a36Mx7csFAgmLT1yBWHtHdCNV5--hsSrZaaGzUHaaa/items/01DNBUO6F3OAAAAAAMKVGIXFKCFDTW7HSH HTTP/1.1
Accept: application/json; odata.metadata=minimal; odata.streaming=true; IEEE754Compatible=true
Accept-Encoding: gzip
User-Agent: NONISV|SharePointPnp|PnPCoreSDK
Content-Type: application/json
Host: graph.microsoft.com
Content-Length: 24
Expect: 100-continue

{"name":"NewFolderName"}

HTTP/1.1 403 Forbidden
Cache-Control: no-store, no-cache
Content-Type: application/json
Vary: Accept-Encoding
Strict-Transport-Security: max-age=31536000
request-id: f9221b1c-496f-4860-afb0-c085b0cf399b
client-request-id: f9221b1c-496f-4860-afb0-c085b0cf399b
x-ms-ags-diagnostic: {"ServerInfo":{"DataCenter":"East US","Slice":"E","Ring":"5","ScaleUnit":"004","RoleInstance":"BL02EPF00021892"}}
Date: Mon, 08 Apr 2024 20:32:12 GMT
Content-Length: 214

{"error":{"code":"accessDenied","message":"Access denied","innerError":{"date":"2024-04-08T20:32:12","request-id":"f9221b1c-496f-4860-afc0-c085b0cf399b","client-request-id":"f9221b1c-496f-4860-afc0-c085b0cf399b"}}}
```

 I'm using an Azure Application Registration for Access and have granted the following recommended permissions:

SharePoint -> Delegated Permissions -> AllSites -> AllSites.FullControl
SharePoint -> Delegated Permissions -> TermStore -> TermStore.ReadWrite.All
SharePoint -> Delegated Permissions -> User -> User.ReadWrite.All
SharePoint -> Application Permissions -> Sites -> Sites.FullControl.All
SharePoint -> Application Permissions -> TermStore -> TermStore.ReadWrite.All
SharePoint -> Application Permissions -> User -> User.ReadWrite.All
Microsoft Graph -> Delegated Permissions -> Group -> Group.ReadWrite.All
Microsoft Graph -> Delegated Permissions -> User -> User.Read
Microsoft Graph -> Delegated Permissions -> Directory -> Directory.ReadWrite.All
Microsoft Graph -> Application Permissions -> Group -> Group.ReadWrite.All
Microsoft Graph -> Application Permissions -> Directory -> Directory.ReadWrite.All

It's a permissions issue.   I don't know what else to check.

P.S.   MoveTo works just fine, and I will be using that until/if the Rename issue is resolved.

**Comment by jansenbe** [comment]
@whobday : you also need the Graph permissions for files/sites. The SharePoint permissions that you selected in Azure AD are only for SharePoint REST and SharePoint CSOM, the SharePoint Graph APIs require files and/or sites permission scopes (https://learn.microsoft.com/en-us/graph/api/driveitem-update?view=graph-rest-1.0&tabs=http).

![image](https://github.com/pnp/pnpcore/assets/7451219/e773892b-30ed-49ed-80a2-87e8e0b0bf49)

**Comment by whobday** [comment]
Thanks for that link; unfortunately, adding those permissions didn't work.  Here is the complete list of permissions for the App Registration(amended with the new permissions):

<img width="655" alt="image" src="https://github.com/pnp/pnpcore/assets/43872402/c57dbb67-9898-4a96-86f1-805a547a1e12">

I will pull down the repo and try to run the tests as detailed [here](https://pnp.github.io/pnpcore/contributing/configuring-for-auth-tests.html). This page was the source of the recommended permissions I had initially applied.

Edited to add:  I'm having the same issue with renaming Libraries and Documents as well.

**Reply by David-Cadmart** [reply]
Hi,

Did you finally find out why was this happening?
I am having the same issue, and I have all the permissions as well.
